src/main/java/com/mkopec/ticketbookingapp/service/SeatValidationService.java

```diff
@@ -0,0 +1,199 @@
+        }
+
+        Integer getNumber() {
+            return roomSeat.getSeat().getNumber();
+        }
+    }
+
+    private static final int NUM_OF_NEIGHBOURS = 2;
+
+    boolean validate(List<RoomSeat> reservedSeats, List<RoomSeat> occupiedSeats, List<RoomSeat> allSeats) throws ArgumentNotValidException {
+        Map<RoomSeat, SeatModel> seats = initializeSeatModels(reservedSeats, occupiedSeats, allSeats);
+
+        // sorting allows to take neighbour of seat
+        Map<String, List<SeatModel>> groupedSeatModelsToRows = seats.values().stream()
+                .sorted(Comparator.comparing(SeatModel::getRow).thenComparing(SeatModel::getNumber))
+                .collect(groupingBy(SeatModel::getRow));
+
+        for (Map.Entry<String, List<SeatModel>> entry : groupedSeatModelsToRows.entrySet()) {
+            List<SeatModel> models = entry.getValue();
+            for (int i = 0; i < models.size(); i++) {
+                SeatModel model = models.get(i);
+                if (model.reserved) {
+                    // check if is possible to reserve
+                    switch (model.seatType) {
+                        case FREE_ON_EDGE: {
+                            model.seatType = SeatType.RESERVATION_DONE;
+                            break;
+                        }
+                        case FREE_INSIDE: {
+                            List<SeatModel> neighbourhood = getNeighbourhoodOfSeat(models, i);
+                            if (isReservationValidForSeatModel(neighbourhood, model)) {
+                                model.seatType = SeatType.RESERVATION_DONE;
+                            } else {
+                                throw new ArgumentNotValidException("Dont left single seat near seat: " + model.roomSeat.getId());
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        return true;
+    }
+
+    private Map<RoomSeat, SeatModel> initializeSeatModels(List<RoomSeat> reservedSeats, List<RoomSeat> occupiedSeats, List<RoomSeat> allSeats) {
+        Map<RoomSeat, SeatModel> seats = new HashMap<>();
+
+        allSeats.forEach(rs -> seats.put(rs, new SeatModel(rs, rs.getIsOnEdge() ? SeatType.FREE_ON_EDGE : SeatType.FREE_INSIDE)));
+        occupiedSeats.forEach(rs -> {
+            if (seats.containsKey(rs)) {
+                seats.put(rs, new SeatModel(rs, SeatType.OCCUPIED));
+            }
+        });
+        reservedSeats.forEach(rs -> {
+            if (seats.containsKey(rs)) {
+                SeatModel seatModel = seats.get(rs);
+                if (!seatModel.seatType.equals(SeatType.OCCUPIED)) {
+                    seatModel.reserved = true;
+                } else {
+                    throw new ArgumentNotValidException("Seat with ID: " + seatModel.roomSeat.getId() + " is already reserved");
+                }
+            }
+        });
+        return seats;
+    }
```

Comment:
dlaczego to nie jest w jakiejś klasie? abstrakcji zamykającej salę kinową?

Reply:
Tak, tutaj można by to opakować w abstrakcje, jakiś typ generyczny implementujący interfejs do pobierania danych zamiast RoomSeat